File: Mar 19 895. Maximum Frequency Stack (Hard)/Solution.go
// Init struct and Constructor
type FreqStack struct {
    freqStack map[int][]int
    valFreq map[int]int
    maxFreq int
}

func Constructor() FreqStack {
    return FreqStack{
		freqStack: make(map[int][]int),
		valFreq:   make(map[int]int),
        maxFreq: 0,
	}
}


func (this *FreqStack) Push(val int)  {
    // Update valFreq
    this.valFreq[val]++
    // Update maxFreq
    currFreq := this.valFreq[val]
    if currFreq > this.maxFreq{
        this.maxFreq = currFreq
    }
    // Append val in freqStack[currFreq]
    this.freqStack[currFreq] = append(this.freqStack[currFreq], val)
}


func (this *FreqStack) Pop() int {
    // Get from maxFreq and last element in freqStack
    lenOfMaxFreq := len(this.freqStack[this.maxFreq])
    popValue := this.freqStack[this.maxFreq][lenOfMaxFreq - 1]
    this.freqStack[this.maxFreq] = this.freqStack[this.maxFreq][:lenOfMaxFreq - 1]
    // Update valFreq
    this.valFreq[popValue]--
    // Update maxFreq
    if len(this.freqStack[this.maxFreq]) == 0{
        this.maxFreq--
    }
    return popValue
}

File: Mar 19 895. Maximum Frequency Stack (Hard)/Solution.py
class FreqStack:

    def __init__(self):
        self.freqStack = defaultdict(list)
        self.valFreq = defaultdict(int)
        self.maxFreq = 0

    def push(self, val: int) -> None:
        # Update valFreq
        self.valFreq[val] += 1
        currFreq = self.valFreq[val]
        # Update self.maxFreq
        self.maxFreq = max(self.maxFreq, currFreq)
        # Append val in currFreq slot
        self.freqStack[currFreq].append(val)
        
    def pop(self) -> int:
        # Get from maxFreq and last element in freqStack
        popValue = self.freqStack[self.maxFreq].pop()
        self.valFreq[popValue] -= 1
        #Update maxFreq
        if not self.freqStack[self.maxFreq]:
            self.maxFreq -= 1
        return popValue
        


# Your FreqStack object will be instantiated and called as such:
# obj = FreqStack()
# obj.push(val)
# param_2 = obj.pop()
